### ISSUE TITLE
Improper selection causes key import to fail with pkcs11 provider for public keys.

### ISSUE THREAD

**Issue by simo5** [openssl/openssl#21596]
OpenSSL is greedy and when the store read keys from providers it tries to import them in the default provider immediately regardless of what operation will happen later.

When an operation later happens that requires to use the provider (in my case ECDH using a private key from the provider) it tries to export the the key and import it back into the pkcs11 provider**.

It does this via evp_pkey_export_to_provider() which unconditionally sets:
int selection = OSSL_KEYMGMT_SELECT_ALL;

This function should probe if the key has a private key via the proper ->has() query and unset the OSSL_KEYMGMT_SELECT_PRIVATE_KEY if a private key is not present.

Otherwise the provider that is being asked to import the key will fail as there are no private parameters in the params passed to the import() function.

Funnily enough this bug went partially undetected because ec_export() completely ignores the fact it has been requested to export a private key and it has none simply omitting to execute the part of the code that export the private point instead of returning an error in key_to_params().

** exporting and re-importing keys is wasteful and bad in various ways as key import is not simple for a token and it causes data loss as information on slot/handle/cka_id etc is lost in the conversion, and should be avoided where possible, but probably material for another issue.

**Comment by simo5** [comment]
I can probably submit a patch to properly set selection in the keymgmt functions, not sure what to do about ec_export() though, it may break code that ignored to properly set selection as "it just worked" for all key types.

**Comment by simo5** [comment]
Ah rsa_export() also does the same silent ignoring ..

**Comment by ifranzki** [comment]
While the documentation is not clear about the selection used with export, it appears that all key management implementations of the default provider behave the way you observed.
Not sure if this is by intention or not....

In doc\internal\man3\evp_keymgmt_util_export_to_provider.pod is sates:

_In all functions that take a selection argument, the selection is used to
constraint the information requested on export. It is also used in the cache
so that key data is guaranteed to contain all the information requested in
the selection._

So the term _is used to constraint_ could be read that way that the selection defines the maximum what is expected to be exported, but not necessarily the minimum. So with that interpretation, the current implementation would be correct. It exports whatever it has available if it falls under the selection. 

Anyway, just note that in the meantime other provides may have/do have adopted that behavior as well. My provider for example (https://github.com/opencryptoki/openssl-ibmca) also follows the current implementation of the default provider export functions. 

So in case you change this, this may cause incompatibilities with other providers....

It may also affect existing applications using `EVP_PKEY_todata` or `EVP_PKEY_export` as they internally also end up calling the providers export function. Applications may rely on the fact that they get whatever is available, and don't check whether the key has a public or private key. There is no `EVP_PKEY_has()` kind of function as far as I can tell, that could be used to check upfront.

I am not arguing that the one or the other way is correct, I am just warning about changing the behavior and its possible consequences....

**Comment by ifranzki** [comment]
The import functions of the default provider first checks what params it finds. Only if it fiends all params required to build the private key, it attempts to import it. So here the selection also seems to be the upper bound of things to import, just like with export. It imports whatever it finds available in the params, but not more than what selection permits. 

Again, I am not arguing if thats the correct way or not, I just see this is how it is done....
At least the selection being the upper bound is consistent across import and export, and also across all algorithms implemented by the default provider.

**Comment by t8m** [comment]
I do not think we can change this - as @ifranzki said it would be almost certainly breaking the compatibility. The documentation should be updated.
Is there any other possible way to fix it?

**Comment by t8m** [comment]
> OpenSSL is greedy and when the store read keys from providers it tries to import them in the default provider immediately regardless of what operation will happen later.

Shouldn't this operation fail for private keys? They aren't usually exportable are they?

**Comment by simo5** [comment]
> I do not think we can change this - as @ifranzki said it would be almost certainly breaking the compatibility. The documentation should be updated. Is there any other possible way to fix it?

I can deal with this, we can turn it into a documentation bug.

**Comment by simo5** [comment]
> > OpenSSL is greedy and when the store read keys from providers it tries to import them in the default provider immediately regardless of what operation will happen later.
> 
> Shouldn't this operation fail for private keys? They aren't usually exportable are they?

This case is default_provider -> pkcs11_provider so any export->import option apply.

However if a key is marked exportable I could also export a pkcs11 provider key out. That said normally the export happens via Key Wrapping, so it is expensive and complicated even when possible (and requires you to provide or receive a wrapping key).

**Comment by simo5** [comment]
One issue here though is that on export, even though I do have a private key, I will not be able to export but a public key, or fail the operation.
If selection does not provide me hints on the purpose I have no way to know if I should fail (because the user really tried to obtain a private key), or I should succeed, because the caller will be content with a public key only. I think selection should, as much as possible, reflect the true intent and not be broad, broad is bad in some situation where an error is better than data loss (think key backup or some such).

**Comment by t8m** [comment]
IMO the selection on export as set by libcrypto should reflect the intention - i.e. it should not set OSSL_KEYMGMT_SELECT_PRIVATE_KEY if only public key is needed. The question is though, whether that "intention" is always known at the time of export. If there are any cases where libcrypto tries to export things it does not need for the subsequent operation, that should be fixed.

And yeah, the provider absolutely must fail the export operation if OSSL_KEYMGMT_SELECT_PRIVATE_KEY is set on export and the key has a private key part which is not exportable.

Default/fips providers do not have such problems because they always have fully exportable keys.

**Comment by t8m** [comment]
> However if a key is marked exportable I could also export a pkcs11 provider key out. That said normally the export happens via Key Wrapping, so it is expensive and complicated even when possible (and requires you to provide or receive a wrapping key).

I am not sure how to handle this case though. One way would be to say that because there is no key wrapping key used in export/import operations that we should treat these keys as unexportable. Or perhaps make that an option of the provider whether exportable private keys should be allowed to be exported under some bogus key wrapping key?
But then once the key is imported into the default provider it should not be re-exported to the PKCS11 provider anymore shouldn't it? Of course unless you made the PKCS11 provider preferred via the propquery but that is not a normal setup.

**Comment by simo5** [comment]
> > However if a key is marked exportable I could also export a pkcs11 provider key out. That said normally the export happens via Key Wrapping, so it is expensive and complicated even when possible (and requires you to provide or receive a wrapping key).
> 
> I am not sure how to handle this case though. One way would be to say that because there is no key wrapping key used in export/import operations that we should treat these keys as unexportable. Or perhaps make that an option of the provider whether exportable private keys should be allowed to be exported under some bogus key wrapping key? But then once the key is imported into the default provider it should not be re-exported to the PKCS11 provider anymore shouldn't it? Of course unless you made the PKCS11 provider preferred via the propquery but that is not a normal setup.

I think we can ignore this problem for now, it is an uncommon thing.
We should:
A) not be greedy in exporting keys from providers at store open time.
B) make sure to set the correct selection when it is clear what the intention or circumstances are.

For A:
I do not understand why the store code is so greedy in trying to export keys to the default provider, so I don't know what can/should be done to avoid needlessly bouncing keys between providers.

For B:
For example when calling evp_keymgmt_util_try_import() the selection should reflect the data the exporter was able to export.
In evp_keymgmt_util_export_to_provider() import_data.selection should be set to what the export was able to achieve, either by calling _has() functions before setting that value, or,  perhaps more accurate, by updating the data->selection just before calling evp_keymgmt_import() after the export operation is performed.

**Comment by t8m** [comment]
The _greediness_ is IMO caused by libcrypto wanting to prefer the provider according to the query. Even empty query prefers a provider - although not documented, it is the one that was loaded first. But once you use a preference query it is clear that the provider preferred by the preference query should take all keys it can use.

So I am not sure we can easily fix that greediness - IMO it is inherent in the existing design and it would probably require incompatible changes to fix it.

B should be possible to be fixed without backwards compatibility problems and IMO it is a bug.

**Comment by simo5** [comment]
> The _greediness_ is IMO caused by libcrypto wanting to prefer the provider according to the query. Even empty query prefers a provider - although not documented, it is the one that was loaded first. But once you use a preference query it is clear that the provider preferred by the preference query should take all keys it can use.

What's the rationale here?
I understand using the preferred provider to try to load a key first, but if a Key is in another provider why prefer the default one? Shouldn't the key stick to the provider of origin until an operation is required on it?
 
> So I am not sure we can easily fix that greediness - it is inherent in the existing design and it would probably require incompatible changes to fix it.

Well, the easy way is to deny export, but has consequences, yet pkcs11 currently deny export from the store provider because of various side-effects I otherwise had, one of them is this bug, but finally I have enough infrastructure to handle it, more or less. pkcs11 still allows export from the keymgmt provider for things like printing out or actually sxporting out to file (but again with some issues I have not figured out).

> B should be possible to be fixed without backwards compatibility problems and it is a bug.

Thanks, is this something I should attempt a patch for? (May take a while)

**Comment by t8m** [comment]
> Thanks, is this something I should attempt a patch for? (May take a while)

If you want to. I was looking at this a few months ago already because of a different issue and it was non-trivial as we are at least in some cases caching the exported data and we must avoid caching it if the export is incomplete (i.e. with selection not including all parameters).

**Comment by simo5** [comment]
> > Thanks, is this something I should attempt a patch for? (May take a while)
> 
> If you want to. I was looking at this a few months ago already because of a different issue and it was non-trivial as we are at least in some cases caching the exported data and we must avoid caching it if the export is incomplete (i.e. with selection not including all parameters).

Yes, this was exactly one of my concerns with incorrect matching between selection and actual data.

**Comment by baentsch** [comment]
Looking for further issues regarding provider import/export documentation, I wound up on this issue and it makes me wonder how this stuff gets its tracking labels:

This issue does not carry "branch: 3.3" and "branch: 3.4" labels, but it also applies to them, right? Also it's not yet fixed for either branch (considering the "branch: master" label), right too?

If so (?), two suggestions if I may: 1) Add a label "branch: 3.x" upon release of "3.x" to all issues tagged "branch: master" at time of release -- should be "automatable" -- but I have no idea how to do that, though: Worthwhile doing? Do you know someone who could help, @t8m?
2) Work on resolving the issue :) Anything anyone already did by now so the issue can simply be closed @simo5 ? Or something maybe I could help with (always wanted to work with other providers :)?

**Comment by t8m** [comment]
> This issue does not carry "branch: 3.3" and "branch: 3.4" labels, but it also applies to them, right? Also it's not yet fixed for either branch (considering the "branch: master" label), right too?

Yes, as this was reported before 3.3 and 3.4 releases it does not get these labels, but given it has branch: master it is implied that 3.3 and 3.4 have the issue as well. We do not automatically add these labels when new branches are created. Although we could with some scripting.

**Comment by t8m** [comment]
> If so (?), two suggestions if I may: 1) Add a label "branch: 3.x" upon release of "3.x" to all issues tagged "branch: master" at time of release -- should be "automatable" -- but I have no idea how to do that, though: Worthwhile doing? Do you know someone who could help, @t8m?

I do not like this - it would be IMO confusing to sometimes see branch: 3.0... branch: 3.4 labels and branch: 3.x elsewhere. IMO better would be to add the intermediate labels by a script. But I do not think it is critically needed.



> 2) Work on resolving the issue :) Anything anyone already did by now so the issue can simply be closed @simo5 ? Or something maybe I could help with (always wanted to work with other providers :)?

Feel free to tackle this, however it is definitely challenging endeavor.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity leading to implementing or re-prioritizing it for more than 2 months.
